cgsmiles/write_cgsmiles.py

```diff
@@ -0,0 +1,223 @@
+            # edge we last crossed to get here is interesting.
+            previous = predecessors[current]
+            assert len(previous) == 1
+            previous = previous[0]
+
+        smiles += format_node(molecule, current)
+        if current in atom_to_ring_idx:
+            # We're going to need to write a ring number
+            ring_idxs = atom_to_ring_idx[current]
+            for ring_idx in ring_idxs:
+                ring_bond = ring_idx_to_bond[ring_idx]
+                if ring_idx not in ring_idx_to_marker:
+                    marker = _get_ring_marker(ring_idx_to_marker.values())
+                    ring_idx_to_marker[ring_idx] = marker
+                    new_marker = True
+                else:
+                    marker = ring_idx_to_marker.pop(ring_idx)
+                    new_marker = False
+
+                smiles += str(marker) if marker < 10 else '%{}'.format(marker)
+
+        if current in dfs_successors:
+            # Proceed to the next node in this branch
+            next_nodes = dfs_successors[current]
+            # ... and if needed, remember to return here later
+            branches.update(next_nodes[1:])
+            to_visit.extend(next_nodes)
+        elif branch_depth:
+            # We're finished with this branch.
+            smiles += ')'
+            branch_depth -= 1
+
+    smiles += ')' * branch_depth
+    return "{" + smiles + "}"
+
+def _find_min_node(molecule, node):
+    fragid = molecule.nodes[node]["fragid"]
+    node_to_fragid = nx.get_node_attributes(molecule, "fragid")
+    return min([node for node, _id in node_to_fragid.items() if _id == fragid ])
+
+def _find_nodes_bonding(molecule, fragname):
+    """
+    For nodes in `graph` check if they participate
+    in any edges that have the `bonding` attribute.
+    If yes store the node and the bonding operator
+    belonging to that node.
+    """
+    edges = nx.get_edge_attributes(molecule, "bonding")
+    node_to_bonding = defaultdict(list)
+    for edge in edges:
+        if molecule.nodes[edge[0]]["fragname"] == fragname:
+            min_node = _find_min_node(molecule, edge[0])
+            node = edge[0] - min_node
+            node_to_bonding[node].append(edges[edge][0])
+        elif molecule.nodes[edge[1]]["fragname"] == fragname:
+            min_node = _find_min_node(molecule, edge[1])
+            node = edge[1] - min_node
+            node_to_bonding[node].append(edges[edge][1])
+    return node_to_bonding
+
+def _smiles_node_iter(smiles_str):
+    organic_subset = 'B C N O P S F Cl Br I * b c n o s p'.split()
+    batom=False
+    for idx, node in enumerate(smiles_str):
+        if node == '[':
+            batom = True
+            start = idx
+
+        if node == ']' and batom:
+            stop = idx+1
+            batom = False
+            yield start, stop
+
+        if node in organic_subset and not batom:
+            yield idx, idx + 1
```

Comment:
This doesn't work, because `node` is a single character (right? Or is the variable name `smiles_str` misleading), but `organic_subset` contains 2-character elements.

Reply:
hmmm yes that is very much true ...